### PR TITLE
Implement checkbox selection

### PR DESCRIPTION
**Pull request by wolmir**
#1924 and #1960 

This PR will add checkboxes for row selection to the Experiments table webview.
It will also allow batch selection of rows by shift-clicking a range of them (see demo)

https://user-images.githubusercontent.com/1231848/176813261-af39d3c3-74c4-4b4d-955a-1e41f5adf638.mov

### TODO

- [x] Add specific tests for the checkboxes and batch selection

### CODE REVIEW

**Comment by wolmir** [comment]
@rogermparent @mattseddon Can you take a look at this, please?

File: webview/src/experiments/components/table/Table.tsx [hunk 1]
```diff
@@ -43,6 +50,7 @@ export const ExperimentGroup: React.FC<RowProp & InstanceProp> = ({
         instance={instance}
         contextMenuDisabled={contextMenuDisabled}
```
Comment (mattseddon) [inline webview/src/experiments/components/table/Table.tsx:51]: Seems like we are still suppressing the full context menu whenever an experiment is running. Perhaps we need to now do this on an action-by-action basis as there is nothing to stop a user from "starring" multiple experiments on the fly whilst an experiment is running.

File: webview/src/experiments/components/App.test.tsx [hunk 2]
```diff
@@ -810,6 +810,40 @@ describe('App', () => {
...
+      const selectedRows = screen.getAllByRole('row', { selected: true })
+      expect(selectedRows.length).toBe(4)
+
+      const target = screen.getByText('4fb124a')
+      fireEvent.contextMenu(target, { bubbles: true })
+
+      jest.advanceTimersByTime(100)
+      const menuitems = screen.getAllByRole('menuitem')
+      const itemLabels = menuitems.map(item => item.textContent)
+      expect(itemLabels).toContain('Star Experiments')
```
Comment (mattseddon) [inline webview/src/experiments/components/App.test.tsx:845]: [C] Could click the button and check the correct message is fired here.

**Review by mattseddon — APPROVED** [review]
A couple of items to follow up on. Will merge this now.

**Comment by mattseddon** [comment]
I would also have at least one row selected and one row starred in the storybook for snapshot testing. You should be able to use the play functionality to achieve that.

File: webview/src/experiments/components/table/Row.tsx [hunk 3]
```diff
@@ -294,11 +299,18 @@ export const RowContent: React.FC<
...
 
-  const toggleRowSelection = React.useCallback(() => {
-    if (!isWorkspace) {
-      toggleRowSelected?.({ row })
-    }
-  }, [row, toggleRowSelected, isWorkspace])
+  const toggleRowSelection = React.useCallback<HandlerFunc<HTMLElement>>(
+    args => {
+      if (!isWorkspace) {
+        if (args?.mouse?.shiftKey) {
```
[the remark this answers is not in the document]
Comment (maxagin) [reply webview/src/experiments/components/table/Row.tsx:305]: > [Q] How do we implement the reverse? After doing a bulk selection I am unable to bulk unselect. Instead, I have to unclick every row manually.

Good point. I have commented this here: https://github.com/iterative/vscode-dvc/issues/1924#issuecomment-1174426634

In the case of multiple selections, should not we have Unselect All option in the menu?

[the remark this answers is not in the document]
Comment (shcheklein) [reply webview/src/experiments/components/table/Row.tsx:305]: It can be Esc?

[the remark this answers is not in the document]
Comment (wolmir) [reply webview/src/experiments/components/table/Row.tsx:305]: I like the Esc option, it's the first key I tried to hit as well.

I discussed an idea with Matt yesterday, but I didn't like it in the end.

**Pull request merged**